server-base/src/main/java/org/apache/kylin/rest/service/QueryService.java

```diff
@@ -708,6 +708,22 @@ private SQLResponse queryWithSqlMassage(SQLRequest sqlRequest) throws Exception
                         borrowPrepareContext = false;
                         preparedContext = createPreparedContext(sqlRequest.getProject(), sqlRequest.getSql());
                     }
+
+                    // when SparkEngine enabled, Execution Plan was created from RelNode directly, RelNode should be cached
+                    if (getConfig().isSparkEngineEnabled()) {
+                        // preparedContext initialized by current thread, put relNode into cache
+                        if (preparedContext.olapRel == null) {
+                            preparedContext.olapRel = QueryContextFacade.current().getOlapRel();
+                            preparedContext.resultType = (QueryContextFacade.current().getResultType());
```

Comment:
Add these code after creating 'PreparedContext' function:
`
} catch (NoSuchElementException noElementException) {
                        borrowPrepareContext = false;
                        preparedContext = createPreparedContext(sqlRequest.getProject(), sqlRequest.getSql());
                        preparedContext.olapRel = QueryContextFacade.current().getOlapRel();
                        preparedContext.resultType = (QueryContextFacade.current().getResultType());
                    }
`

Reply:
NoSuchElementException will no be thrown, because PreparedContextFactory will create one